### PR TITLE
Require an initial click on embed previews for interactivity

### PR DESCRIPTION
**Pull request by notnownikki**
## Description

This fixes the block selection UX for video embeds. Previously
the click on the video to select the block would start the
video playing. Now that previews require a click to become
interactive, the initial click selects the block without
side effects.

## How has this been tested?

Embed a youtube video. Click outside of the block. Click the video. The block should be selected, and the video should not start to play.

## Types of changes
Bug fix

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by notnownikki** [comment]
This fixes more issues too, such as if you embed a twitter timeline, you can get caught scrolling the timeline as you use the mousewheel to scroll down the page. But with this, you scroll as expected.

File: packages/block-library/src/embed/editor.scss [hunk 1]
```diff
@@ -31,3 +31,13 @@
 		word-break: break-word;
 	}
 }
+
+.wp-block-embed-interactive-overlay {
+	position: absolute;
+	width: 100%;
+	height: 100%;
```
Comment (afercia) [inline packages/block-library/src/embed/editor.scss]: when using top, right, bottom, left, there's no need for width and height.

File: packages/block-library/src/embed/embed-preview.js [hunk 2]
```diff
@@ -17,57 +17,93 @@ import classnames from 'classnames/dedupe';
...
+					html={ html }
+					scripts={ scripts }
+					title={ iframeTitle }
+					type={ sandboxClassnames }
 				/>
-			) }
-		</figure>
-	);
+				{ ! interactive && <div
+					role="dialog"
```
Comment (afercia) [inline packages/block-library/src/embed/embed-preview.js]: this div is not an ARIA dialog, doesn't have a particular semantic meaning nor can be focused so this role needs to be removed

**Comment by afercia** [comment]
Left a couple minor comments. More importantly, seems to me while this approach helps with selecting the block, it also introduces a potentially confusing interaction model,

Especially users who use input devices other than mouse / touch or assistive technologies might be confused. Also, I'm not sure mouse / touch users would understand why a first click doesn't "play" and a second click does. I'd tend to think we'd need a different approach.

We should always take into consideration all devices and design for all users. Keyboard users, for example, can still get to the player controls: in the screenshot below, I've used a rgba red background on the overlay to make it clearly visible:

screenshot 2018-12-19 at 17 21 19

As you can see, my mouse is on the paragraph above the embed. With the keyboard, I'm still able to tab through all the player controls and activate them (see the "watch later" control). Many users make a mixed use of mouse and keyboard: most likely, they would be confused by the different interaction with the overlay.

If we really want to make the player "inert" to allow a better block selection mechanism, then I'd tend to think the player should be "inert" for _all users_ until the block is selected. Offering a different experience to different users seems to me not so ideal. 

A while ago, there has been an experiment to introduce a new selection and navigation mechanism with the blocks: see "navigation and edit mode" tried in #5709 and recently revived with the proposal in #11581. I'd be in favor of pairing the experience for all users: navigation / edit mode would help users to have the same experience, regardless of the device they use.

**Comment by notnownikki** [comment]
> If we really want to make the player "inert" to allow a better block selection mechanism, then I'd tend to think the player should be "inert" for all users until the block is selected. Offering a different experience to different users seems to me not so ideal.

That's not possible, unfortunately. If we have an overlay making previews inert based on `isSelected`, then mouse clicks get through to the preview when selecting the block because as soon as the mouse goes down, the block becomes selected and the overlay is switched off, and the click goes through to the underlying preview. It's a really horrible problem, but IMO having a slightly inconsistent experience for combined keyboard/mouse users is better than having videos that play when you try to select a block with a touch or click.

**Comment by jasmussen** [comment]
I'm about to head out for some family things, but I have ideas to share that could get a version of this in. We also have some past mockups that I think we can dig up. I think there's a path forward, but I'll share more tomorrow.

**Comment by afercia** [comment]
> That's not possible, unfortunately

It is possible, at least for keyboard users. I've pointed to #5709 / #11581 have you had a look at those?

Screen reader users would need something more, as with a screen reader it's always possible to use arrows navigation to get the content.

> having a slightly inconsistent experience for combined keyboard/mouse users is better than ...

I disagree 🙂We're aiming to build accessible software. The first step is providing the same content and experience to all users in a device-independent way, as this is the real nature of the web.

**Comment by jasmussen** [comment]
Thank you for fearlessly taking a stab at this. I strongly agree this is something we need to fix. 

Regarding the specific way we fix this, I have reopened #483 which has a bunch of mockups, and was supposedly fixed in #5730. Do we know why that fix regressed? CC: @aduth 

In any case, to address the mouse/keyboard situation, can the mockup on that old ticket inspire an approach that would be the same for both?

click blocker

I imagine it could work like this:

1. You load up an existing post with a video embed. No block is selected. You see the video embed clearly — it is not dimmed in any way.
2. You select the video embed. Whether through keyboard navigation or a click. You see the dimmed version above, and the video does not start by being selected. 

Is the above possible? Does it address the keyboard/mouse issues?

If we need for you to be able to interact with the contents of the embed inside the editor, we could add a a button on top of the dimmed/disabled embed: "Interact with content" (or a better label). Clicking this button would make the embed not disabled again, until you deselect the block. It could look like this:

dimmer-mockup

Can we use tabIndex to make the embed container un-interactable by keyboard until it has been "unlocked" with the button press?

**Comment by notnownikki** [comment]
> Can we use tabIndex to make the embed container un-interactable by keyboard until it has been "unlocked" with the button press?

This is possible as long as we change the `tabindex` directly on the DOM node instead of changing the `tabindex` prop. If we change the prop, the iframe is re-rendered, and the preview is regenerated and the UI jumps about. So yeah, it's possible, the code just needs to have plenty of comments explaining why it's not being done in the obvious way :)

> If we need for you to be able to interact with the contents of the embed inside the editor, we could add a a button on top of the dimmed/disabled embed: "Interact with content" (or a better label).

As I understand it, it would need to be done this way with an interactive element for screen readers to work properly. One of the issues brought up in the core editor chat was that the static `div` used in the current code would not work with screen readers.

**Comment by jasmussen** [comment]
> As I understand it, it would need to be done this way with an interactive element for screen readers to work properly. One of the issues brought up in the core editor chat was that the static div used in the current code would not work with screen readers.

But a static div with a button inside, clearly labeled, perhaps even with a verbose description, that would presumably be fine, right? 

I think we need to lean into the design principle that a _selected block_ can show additional controls and interface. So I think this shouldn't be a problem.

**Comment by notnownikki** [comment]
> But a static div with a button inside, clearly labeled, perhaps even with a verbose description, that would presumably be fine, right?

Yes, absolutely. The problem with this PR is it's _only_ a static div, no elements that a screenreader would recognise. So the second mockup would be great!

However, if we got the `FocusableIframe` fixed, would we want this anyway?

**Comment by jasmussen** [comment]
> However, if we got the FocusableIframe fixed, would we want this anyway?

Ideally yes, even if not as part of this PR (and by that I mean I think it's okay to ship a small fix and then a bigger fix later). But I have a feeling this PR might need to go into 5.1 as opposed to 5.0.3 which might not happen depending on things. 

But yes, because the way it is right now, any YouTube videos appear "interactable" even if they aren't (i.e. they require two actions). It's probably better to embrace that we are intercepting the actions and making it more visually clear what's happening.

The balance in all these cases is finding a way that isn't creating "heavy UI", so as to consider all aspects of accessibility, including use cases 1 and 6 in [this poster](https://accessibility.blog.gov.uk/wp-content/uploads/sites/52/2016/08/Karwai-blog-1-1.png). But because this "action interceptor UI" is only visible when the block is selected, it is not additional UI that weighs down the whole thing, and it's clear how to "get it away" (with the button action).

**Comment by afercia** [comment]
> Can we use tabIndex 

Short version: no 🙂 

Long version: assistive technologies do allow to navigate to elements with a tabindex -1 so the embed content would still be interactive for these users. As accessibility team we should try to explain this better to all the designers involved, as it seems to be a recurring question.

Looking at the history, I see in in https://github.com/WordPress/gutenberg/pull/517#discussion_r113538380 etc. there is also some discussion about the different embed _types_. Should this apply to only videos or to all the embed types?

I see the overlay approach was already implemented and then removed in #816. Seems to me the main concerns were about scopint the CSS https://github.com/WordPress/gutenberg/pull/816#issuecomment-302054930 and problems with a "resizing iframe" https://github.com/WordPress/gutenberg/pull/816#issuecomment-305130590

Later, a solution that used the blur event was implemented in https://github.com/WordPress/gutenberg/pull/4011 to  my understanding that's no longer possible after #4872.

Lastly, the issue was fixed again in https://github.com/WordPress/gutenberg/pull/5730 as Joen mentioned. Checked in Gutenberg 2.9.0 it worked. Worth noting the video couldn't be played at all, not even after multiple clicks. Stopped working in Gutenberg 3.0.0 so maybe this part is a regression?

Granted, even in 2.9.0 keyboard and assistive technologies users were still able to tab to the video player so even then the implemented solution was mouse-only.

Trying some lateral thinking: in #483 Joen mentioned "Abstract live preview". Do we really need to display a fully functional video? Is it possible to display the video poster image instead? Do all video embed providers have a poster image in their response? This would also make the page way lighter. A button https://github.com/WordPress/gutenberg/issues/483#issuecomment-297280468 could be used to load the actual video + player.

**Comment by jasmussen** [comment]
> Do we really need to display a fully functional video?

So long as these embeds are super basic (and they really are just the oembeds at this point), I don't personally think they should be fully functional. Though I would really hate for us to go into the abstracted live preview territory, even if I did think so in the past. With the improvements made to editor styles, the editor itself is increasingly a preview of the final content, and discrepancies in that preview are therefore likely to cause user concerns: "Am I doing this right? It looks wrong". 

This used to be solved, and I know this is solved in the classic editor. I think Ella is AFK, but perhaps @azaozz can provide insights?

For what it's worth, if we can make the entire embed completely inert, for mouse and keyboard users alike — totally un-interactable — I think that would be fine, so long as the visual of the embed is the same. And if we did this, yes, it would be nice if we could do this for only videos, so non video embeds were still interactible.

**Comment by notnownikki** [comment]
> Short version: no ￼

:sob: 

> Later, a solution that used the blur event was implemented in #4011 to my understanding that's no longer possible after #4872.

:sob: Ok, then we should remove the `FocusableIframe` from the embed block altogether. That explains why the debug code I tried didn't work, if the focus events are no longer supported.

> Trying some lateral thinking: in #483 Joen mentioned "Abstract live preview". Do we really need to display a fully functional video? Is it possible to display the video poster image instead?

:bell: :bell: :bell: :bell: I think we might have a winner here :grin: Allowing an email to show the poster with a button which displays an interactive full preview when clicked sounds like it sovles all of this.

@jasmussen :

>  I don't personally think they should be fully functional. 

Sorry, but they really should be for the next round of embed enhancements that change things like YouTube start time, player theme, branding, Twitter media options, etc.... It would be a poor UX if we had to get the user to preview their post to check all the options were set right and looked right, when the current implementation of that branch (admittedly, only on my local machine right now...) updates the preview live.

**Comment by notnownikki** [comment]
> This used to be solved, and I know this is solved in the classic editor.

Looks like it used the same approach as we had before with the FocusableIframe, which we can't use now because of the API change.

**Comment by afercia** [comment]
> I know this is solved in the classic editor

Yep I think it's a TinyMCE thing. However, even in the Classic Editor users are able to get to the player when using a screen reader... it's just a bit more difficult.

screenshot 2018-12-20 at 10 26 34

> For what it's worth, if we can make the entire embed completely inert, for mouse and keyboard users alike — totally un-interactable — I think that would be fine, so long as the visual of the embed is the same.

Yeah I'd love to see some exploration in that direction.

> Ok, then we should remove the FocusableIframe from the embed

Not sure #4011 / #4872 have something to do with that 🙂 there must be something else that makes the  not work as intended.

**Comment by jasmussen** [comment]
> Sorry, but they really should be for the next round of embed enhancements that change things like YouTube start time, player theme, branding, Twitter media options, etc.... It would be a poor UX if we had to get the user to preview their post to check all the options were set right and looked right, when the current implementation of that branch (admittedly, only on my local machine right now...) updates the preview live.

I agree, but if step 1 was to make the embeds completely inert, step 2 could then be to introduce the dimmer + button from the mockup to make it not inert, would that work?

**Comment by notnownikki** [comment]
> I agree, but if step 1 was to make the embeds completely inert, step 2 could then be to introduce the dimmer + button from the mockup to make it not inert, would that work?

If you do step 1 by itself, you get rid of embed previews completely for things like Twitter that have no thumbnail. Please don't do that :smile:

**Comment by jasmussen** [comment]
> If you do step 1 by itself, you get rid of embed previews completely for things like Twitter that have no thumbnail. Please don't do that 😄

Well, I'm not referring to the abstracted preview that Andrea suggested, I'm not in love with that idea. I'm suggesting we still have the embed exactly as it is in this very branch, but it just utterly inert, you can click it to select the block, but nothing inside the iframe is interactible. I'm not sure it's possible, clearly we can't redirect the tabindex to let you not select it, but I know that for clicks we can use `pointer-events: none;` — perhaps we need to use JS to let you not tab into it?

**Comment by notnownikki** [comment]
I would rather do step 1 and 2 at the time time. For embeds that have a static representation supplied by oEmbed, we show that with a button. When you click the button, you get a live preview. When the block is no longer selected, we switch back to the static representation. That's pretty much what this branch does, but instead of "static representation", read "click intercepting overlay"

**Comment by notnownikki** [comment]
Just to clarify - doing step 1 and 2 at the same time could be done in a way that only affects the video embeds. For the options work, we inject components to affect the preview, so there's no reason that video embeds couldn't use a `VideoEmbedPreview` component that does the thumbnail+button -> live preview switching, while everything else carries on unaffected.

**Comment by notnownikki** [comment]
I think to start with, I'll look at fixing `FocusableIframe` so it works as it used to, then the need for overlays, buttons, etc. goes away. I'm strongly of the opinion that having a preview that's something other than the actual content is a step backwards, so if we can have it seamless like it used to be, then that would be the best solution.

**Comment by azaozz** [comment]
> > I know this is solved in the classic editor

> Yep I think it's a TinyMCE thing.

Nope, it's a WordPress thing. It's the `wpview` we did with Ella ~three years ago :)
(It's currently broken in the Classic Block because of some major CSS regressions.)

Right. This is solved in the classic editor and not having it in Gutenberg would be bad. As far as I see this PR implements the same solution: adds an overlay that renders the embed "inert" and is removed/hidden on click. That seems the best user experience for interactive embeds in an editor. Since we don't know what the embed is, all embeds should be considered interactive (the same way they all should be in an iframe to isolate them from the main window).

> As you can see, my mouse is on the paragraph above the embed. With the keyboard, I'm still able to tab through all the player controls and activate them (see the "watch later" control). Many users make a mixed use of mouse and keyboard: most likely, they would be confused by the different interaction with the overlay.

Looking at that screenshot, the mouse cursor seems to be hovering over the above block/paragraph, but the actual selected block is the video. Isn't that how it always works? The cursor is in one spot but you can move (by typing or by tabbing) to other places/elements on the page and focus them? Not sure how that is confusing :) 

> One of the issues brought up in the core editor chat was that the static div used in the current code would not work with screen readers.

Hmm, how is that different from the classic editor and why it is not a problem there?

IMHO:
1. All embeds should be considered interactive content (like links).
2. It is great to prevent interaction inside the editor, and allow it "on demand". Just like clicking on a link doesn't navigate away or open a new page.

**Comment by notnownikki** [comment]
@azaozz thank you :)

I'm working on a solution that has the overlay handled by the sandboxed document, so we can handle all the focus / blur events and have the same experience for keyboard, mouse, and screenreader.

**Comment by azaozz** [comment]
Uh, a third iframe? What if the overlay is inside the top iframe (that we add)? Then it can be handled by a bit of js that we add indie of it. I.e. we overlay the body of the iframe and toggle that on click/focus, just like a standard "modal overlay"?

**Comment by notnownikki** [comment]
> What if the overlay is inside the top iframe (that we add)? Then it can be handled by a bit of js that we add indie of it. I.e. we overlay the body of the iframe and toggle that on click/focus, just like a standard "modal overlay"?

Pretty much what I'm doing :) 

> Hmm, we will need to pass the parent's (the block) focus state to enable the overlay, but that seems doable as the iframe is "local".

That doesn't work, because the inital mouse click sets the block to focused, and the overlay never renders and the click goes to the underlying preview.

> Thinking more about this... Third iframe may be the easier solution as we can control it "the standard way" through the APIs.

So what I'm working on now is injecting the overlay into the sandboxed HTML. Then we're not worrying about how G handles focus or selected status. We just have to have some elements in the sandboxed document that pick up focus resulting from keyboard navigation, and listen for touches and mouse clicks.

**Comment by azaozz** [comment]
> > Hmm, we will need to pass the parent's (the block) focus state to enable the overlay, but that seems doable as the iframe is "local".

> That doesn't work, because the inital mouse click sets the block to focused, and the overlay never renders and the click goes to the underlying preview.

Hmm, it should work when:
1. The overlay is enabled initially.
2. We let a click or focus of the iframe content remove it (and expose the inner iframe with the actual embed).
3. We listen to the block's blur, pass it inside the iframe and enable the overlay again.

**Comment by notnownikki** [comment]
> We listen to the block's blur, pass it inside the iframe and enable the overlay again.

Pretty much exactly what this PR does :) When `isSelected` becomes false, the overlay appears again. The active state of the overlay is managed as internal state, instead of depending on `isSelected`, but a `false` `isSelected` always resets it to active.

The issue is with keyboard navigation, which seems easier to fix by moving the overlay inside the sandbox, but I'm still experimenting to see which is cleaner.

**Comment by notnownikki** [comment]
For keyboard navigation to remove the overlay, we have to have dummy `a` elements at the top and bottom of the preview, because if the only focusable element in the document is an iframe, then we don't get notified of the focus change and we don't get to remove the overlay when we tab into the preview iframe. I'll try all the approaches and see which of them A) work, and B) aren't horrible to maintain :smile:

**Comment by notnownikki** [comment]
@afercia Looking at keyboard and screen reader navigation here. I appreciate this might not be a long term solution with the other approaches we're looking at, and at the same time the current UX for selecting embedded videos is terrible, so I'm trying to come up with something that works while we look at longer term solutions.

This PR works for mouse and touch users, but keyboard users can access the video without the overlay getting disabled.

The problem is that if you navigate to the preview content using the keyboard, the focus events go to the iframe. We don't find out that it's been navigated to. And if we turn off the overlay when `isSelected` becomes `true`, then mouse clicks get through because `isSelected` becomes `true` as soon as the mouse goes down and it propagates down to the preview, which then receives the click.

So what I have working is something like this:

```
<a />
< ... preview content ... >
<a />
```

The two `a` elements serve to catch keyboard navigation so we get a focus event when the user is navigating to the preview, so the overlay gets removed and the experience is consistent. Without a large change in navigation and selection as the other issues and PRs propose, I'm not sure what more we can do in the short term.

**Comment by notnownikki** [comment]
Final update for the day... probably ;)

Looks like we can have a unified experience, because `FocusableIframe` _is_ working and we can use the same approach as in this PR, and make it work for screenreader navigation and keyboard navigation by having the `onFocus` on the iframe remove the overlay.

I'm not sure what element to use for the overlay to be most friendly to screenreaders, but I think that's something that can be changed once the navigation issues are sorted.

**Comment by afercia** [comment]
@azaozz thanks :) re: the screenshot, I've mentioned the mouse is hovering the paragraph above just to make clear I wasn't using it an, insted, I was using the keyboard on the embed block. So the problem is not that one.

> Not sure how that is confusing

It is because when using the mouse, the embed block works in a way, and when using the keyboard it works in another way.  Many users make a mixed use of mouse / keyboard so... 

> > One of the issues brought up in the core editor chat was that the static div used in the current code would not work with screen readers.

> Hmm, how is that different from the classic editor and why it is not a problem there?

I haven't said is not a problem there. It is. However, we're working on a different editor now seems to me this would be a good opportunity to implement something better compare to the past: a unique, predictable interaction that works in the same way for _all users_. Regardless of the their device, software, ability.

**Comment by azaozz** [comment]
> It is because when using the mouse, the embed block works in a way, and when using the keyboard it works in another way.

Yep, exactly. But mouse and keyboard are different "devices" that work differently. IMHO we should be trying to create the best user experience for the users of each device, not just "copy" how one device works and "paste" it to the other :)

I agree, in most cases we need to try to match how both mouse and keyboard work for "navigation" of the screen. Perhaps this case with the overlay is an exception. 

- Mouse users would typically click in the middle of the block to select it. This is the expected behavior for all blocks. If there is no overlay that may or may not select the block (depending on the nested iframes in it) and will "trigger" the embed's controls, i.e. if there is a video in an iframe the video will start playing but the block may not be "selected", no toolbar, etc. This is the main annoyance that we're trying to prevent. When clicking on an embed the first click should select the block and make the embed "active", then a second click would go through to the embedded content and let the user interact with it.
- Keyboard users don't have the same problem/annoyance. They cannot "click" in the middle of the block. They will use tab or arrow keys to reach the block first. That will "select" the block. Then they will continue to the block's content. If there is another, hidden, "tab stop" in there that prevents them from reaching the embedded content, it will actually be an annoyance for them and may be a bit confusing.

**Comment by notnownikki** [comment]
Gotcha. I'll look at it again.

**Comment by afercia** [comment]
> But mouse and keyboard are different "devices" that work differently. IMHO we should be trying to create the best user experience for the users of each device,

🙂one of my points is that there's no black and white: a lot of users make a mixed use of mouse and keyboard. For example, consider this scenario:
- using the keyboard, I can tab through the player UI and play the video
- the overlay is still there though
- then I want to stop it and I try to use the mouse
- when I click, nothing happens because the overlay is still there
- I'll probably try to click again and yes, this time the video stops
To me, this would be confusing as I wouldn't know why the first click didn't work. And it's just one possible scenario. Taking into consideration also alternative devices, screen readers, etc. I'd tend to think the user experience could be even more confusing.

Also, the root cause of this issue is a design issue: the UI doesn't have a good block selection mechanism when the block content is interactive. The design should provide such mechanism. Instead, we're trying to solve this issue through code, making a `div` That said, ideally the user experience should be the same for all users, regardless of the device they use.

I wish there was a way to make an interactive embed (any embed, also tweets for example) completely "inert" for all users. Not sure that's possible. That said, fair enough: the overlay is an improvement for some users. What about #5730 though? Still not clear why it doesn't work any longer. Also, should the `focusable-iframe` be removed now that the overlay is going to be used?

**Comment by azaozz** [comment]
> I'll probably try to click again and yes, this time the video stops
To me, this would be confusing as I wouldn't know why the first click didn't work. And it's just one possible scenario. Taking into consideration also alternative devices, screen readers, etc. I'd tend to think the user experience could be even more confusing.

Would that be more confusing than having a "tab stop" that does nothing? I also use keyboard + mouse "page navigation" sometimes (not just for testing) and my expectation is that I have to click somewhere on the screen before I can switch to keyboard navigation (as the cursor/selected element is usually somewhere else). Similarly, switching from keyboard to mouse navigation is another step. That seems to work pretty well in the classic editor now.

> the root cause of this issue is a design issue

Right. It is, at the root, a browser issue. Unfortunately the only way we can fix it is by "taking control" away from the browser and adding something better with js :)

**Comment by notnownikki** [comment]
@afercia `focusable-iframe` may be the saviour here :) While it can't respond to the initial block focus in the way we'd like, it _can_ remove the overlay when navigated to with the keyboard. Which means the UI will be in the state a combined mouse/keyboard user (that's also me, btw!) would expect. I'm working on a clean version of the code for this today.

**Comment by notnownikki** [comment]
Hi all, new commit pushed up that I think addresses the keyboard and mouse navigation issues.

The result is that after the block gets focus, the preview becomes interactive. After you click to select it, the overlay is removed. If you navigate to it using the keyboard, the overlay is removed, thanks to a new `onFocus` prop on `SandBox`.

**Comment by afercia** [comment]
@notnownikki thanks for all your work and explorations here 🙂

**Comment by afercia** [comment]
P.S. just pulled latest changes, I don't seem to have any overlay 🙂

**Comment by notnownikki** [comment]
@afercia I just did exactly the same. Spent the last 20 minutes rebuilding and wondering why the overlay wasn't there :laughing: 

Anyway, I pushed up a small change that adds a color to the overlay so we can see more easily when it's there or not to evaluate if it's working in the way we want. Here's a gif showing mouse and keyboard interaction.

overlay

**Comment by gziolo** [comment]
@notnownikki - what's the current status of this one? I moved it to 5.1 Gutenberg milestone as it doesn't seem to be production ready. I personally would love this change included in Gutenebrg as soon as possible because it's very annoying when you click on the embed and it starts playing when your intent was to select it.

**Comment by notnownikki** [comment]
The UX needs review so we know the approach is good, and then I'll get the conflicts resolved and the code production ready.

@afercia would you be able to give this another look? I think we've got something we'll be happy with now, but I would like your review :)

**Comment by jasmussen** [comment]
This is working well for me, and the experience is the same for mouse:

embeds

and keyboard:

tabbing

Next step is probably a rebase, and changing the overlay to match something like the proposal here: https://github.com/WordPress/gutenberg/pull/12981#issuecomment-448919552

Let's get this shipped! Nice work Nikki!

**Comment by notnownikki** [comment]
Oh the red overlay was just to make it absolutely clear when it's active and when it's not during testing :) With a totally transparent overlay, it should be fine too? I'll rebase and make it totally transparent and we can test and see what we think. I feel that changing the overlay should be in a different PR, as I would like some more discussion on that, if that's ok?

**Comment by jasmussen** [comment]
> With a totally transparent overlay, it should be fine too?

Interesting. I'm not sure — definitely worth testing — yes that might work.

**Comment by notnownikki** [comment]
Ready for review! :)

**Review by aduth — CHANGES_REQUESTED** [review]

**Comment by notnownikki** [comment]
@aduth thanks for catching those things, they should all be addressed now

File: packages/block-library/src/embed/embed-preview.js [hunk 3]
```diff
@@ -17,58 +17,96 @@ import classnames from 'classnames/dedupe';
...
+		const html = 'photo' === type ? getPhotoHtml( preview ) : preview.html;
+		const parsedHost = parse( url ).host.split( '.' );
+		const parsedHostBaseUrl = parsedHost.splice( parsedHost.length - 2, parsedHost.length - 1 ).join( '.' );
+		const cannotPreview = includes( HOSTS_NO_PREVIEWS, parsedHostBaseUrl );
+		// translators: %s: host providing embed content e.g: www.youtube.com
+		const iframeTitle = sprintf( __( 'Embedded content from %s' ), parsedHostBaseUrl );
+		const sandboxClassnames = classnames( type, className, 'wp-block-embed__wrapper' );
+
+		/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
+		/* eslint-disable jsx-a11y/no-static-element-interactions */
```
[the remark this answers is not in the document]
Comment (notnownikki) [reply packages/block-library/src/embed/embed-preview.js]: Yeah, just for readability :)

**Review by mapk — APPROVED** [review]
I really like that the direction went toward a transparent overlay. From a design standpoint, this works wonderfully. It felt natural to click the block, and then click the embed to get it to play.

File: packages/block-library/src/embed/embed-preview.js [hunk 4]
```diff
@@ -17,58 +17,96 @@ import classnames from 'classnames/dedupe';
...
 
-	const embedWrapper = 'wp-embed' === type ? (
-		<WpEmbedPreview
-			html={ html }
-		/>
-	) : (
-		<div className="wp-block-embed__wrapper">
-			<SandBox
+	static getDerivedStateFromProps( nextProps ) {
+		if ( ! nextProps.isSelected ) {
```
[the remark this answers is not in the document]
Comment (aduth) [reply packages/block-library/src/embed/embed-preview.js]: The two points of this preceding comment were not addressed.

I've verified that a warning would be logged by allowing this function to return `undefined` (i.e. when there is no explicit `return` value).

```
EmbedPreview.getDerivedStateFromProps(): A valid state object (or null) must be returned. You have returned undefined.
```

Of my own curiosity, I dug deeper to see the work involved in generating a new state value for every render of the component. At the very least, it involves a wasteful shallow clone of the state object:

https://github.com/facebook/react/blob/f2e2637c8e351611dc012bba646431feb250a3e8/packages/react-reconciler/src/ReactFiberClassComponent.js#L170

Instead, if the intention is that state only needs to update if the block is both not selected, but has the interactive state, this ought to be expressed as:

```js
static getDerivedStateFromProps( props, state ) {
	if ( ! props.isSelected && state.interactive ) {
		// ...
		return { interactive: false };
	}

	return null;
}
```

This ensures that we only update state when `state.interactive` was already `true`,

**Review by aduth — CHANGES_REQUESTED** [review]
See: https://github.com/WordPress/gutenberg/pull/12981#discussion_r254393032

File: packages/block-library/src/embed/embed-preview.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (notnownikki) [reply packages/block-library/src/embed/embed-preview.js]: Ah, apologies, this one didn't make it on to the todo list. WIll address and update.

**Comment by designsimply** [comment]
Quick note to say this should fix #7872 (and I'm happy loop back to help test once this is ready if you ping me!).

File: packages/block-library/src/embed/embed-preview.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (notnownikki) [reply packages/block-library/src/embed/embed-preview.js]: This should be correct now. Thank you :)

**Comment by paaljoachim** [comment]
What about just a smaller hit area to begin playing the video?

**Review by aduth — APPROVED** [review]
One final question to consider:

Do we really need any of the focus/click handling or `interactive` state, or would it have been enough to just show the overlay when the `isSelected` prop is `false`? In some brief testing, this seems to me to have the exact same effect as what's proposed here.

In other words, I wonder if it would have been enough for the changes to `embed-preview.js` to amount to:

```diff
diff --git a/packages/block-library/src/embed/embed-preview.js b/packages/block-library/src/embed/embed-preview.js
index efc11b7a5..08e29dff4 100644
--- a/packages/block-library/src/embed/embed-preview.js
+++ b/packages/block-library/src/embed/embed-preview.js
@@ -47,6 +47,9 @@ const EmbedPreview = ( props ) => {
 				title={ iframeTitle }
 				type={ sandboxClassnames }
 			/>
+			{ ! isSelected && (
+				<div className="block-library-embed__interactive-overlay" />
+			) }
 		</div>
 	);
```

**Comment by notnownikki** [comment]
The reason for the state was because `isSelected` becomes true on mouse down, and because the mouse down event goes to the overlay but the mouse up doesnt, the block state gets confused and we end up entering multi selection mode with the mouse up. That may have changed, or be browser specific? But that was the reason for needing the state at time of implementation. Try putting an embed block in the middles of a load of paragraphs, clicking it, and then moving the mouse around. If it's not multi selecting, then whatever was causing that has gone away and I'll look at it further when im back from afk on Monday 🙂

**Comment by aduth** [comment]
Ah, yes! And my memory must be failing me, since I had just [remarked the other day](https://github.com/WordPress/gutenberg/pull/12981#discussion_r253659742) on the specific comment describing this behavior 😅

**Comment by notnownikki** [comment]
Thanks to everyone for the help on this one!

**Pull request merged**